core/trino-main/src/main/java/io/trino/execution/scheduler/FixedCountNodeAllocatorService.java

```diff
@@ -0,0 +1,274 @@
+    {
+        FixedCountNodeAllocator allocator = new FixedCountNodeAllocator(session, maximumAllocationsPerNode);
+        allocators.add(allocator);
+        return allocator;
+    }
+
+    private class FixedCountNodeAllocator
+            implements NodeAllocator
+    {
+        private final Session session;
+        private final int maximumAllocationsPerNode;
+
+        @GuardedBy("this")
+        private final Map<Optional<CatalogName>, NodeSelector> nodeSelectorCache = new HashMap<>();
+
+        @GuardedBy("this")
+        private final Map<InternalNode, Integer> allocationCountMap = new HashMap<>();
+
+        @GuardedBy("this")
+        private final LinkedList<PendingAcquire> pendingAcquires = new LinkedList<>();
+
+        public FixedCountNodeAllocator(
+                Session session,
+                int maximumAllocationsPerNode)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.maximumAllocationsPerNode = maximumAllocationsPerNode;
+        }
+
+        @Override
+        public synchronized ListenableFuture<InternalNode> acquire(NodeRequirements requirements)
+        {
+            try {
+                Optional<InternalNode> node = tryAcquireNode(requirements);
+                if (node.isPresent()) {
+                    return immediateFuture(node.get());
+                }
+            }
+            catch (RuntimeException e) {
+                return immediateFailedFuture(e);
+            }
+
+            SettableFuture<InternalNode> future = SettableFuture.create();
+            PendingAcquire pendingAcquire = new PendingAcquire(requirements, future);
+            pendingAcquires.add(pendingAcquire);
+
+            return future;
+        }
+
+        @Override
+        public void release(InternalNode node)
+        {
+            releaseNodeInternal(node);
+            processPendingAcquires();
+        }
+
+        public void updateNodes()
+        {
+            processPendingAcquires();
+        }
+
+        private synchronized Optional<InternalNode> tryAcquireNode(NodeRequirements requirements)
+        {
+            NodeSelector nodeSelector = nodeSelectorCache.computeIfAbsent(requirements.getCatalogName(), catalogName -> nodeScheduler.createNodeSelector(session, catalogName));
+
+            List<InternalNode> nodes = nodeSelector.allNodes();
+            if (nodes.isEmpty()) {
+                throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+            }
+
+            List<InternalNode> nodesMatchingRequirements = nodes.stream()
+                    .filter(node -> requirements.getAddresses().isEmpty() || requirements.getAddresses().contains(node.getHostAndPort()))
+                    .collect(toImmutableList());
+
+            if (nodesMatchingRequirements.isEmpty()) {
+                throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
```

Comment:
That is a good question. I mimicked what `FixedCountNodeAllocatorService` does. I think the motivation for that is to catch a corner case when cluster is misconfigured and query would never complete, e.g. because all the nodes which are registering into cluster are missing catalog a query needs.
Maybe we should not care and just wait for the node to appear; risking it would never happen. @arhimondr opinion?

Reply:
Can it happen that cluster will start accepting queries (because coordinator initialized) and they will fail because workers didn't join yet? Maybe it's not an issue really.

E.g. we already have:
```
query-manager.required-workers
query-manager.required-workers-max-wait
```

How does it play with this? Should we have a timeout instead?

Reply:
`query-manager.required-workers` does not take care task requirements so there is still a chance that we will go pass that and get immediate failure on `NodeAllocator`.
Still I do not see it as a major limitation. We can change immediate failure to wait-with-timeout, but I would do that as a followup. I will issue a ticket for that.

Reply:
https://github.com/trinodb/trino/issues/11020